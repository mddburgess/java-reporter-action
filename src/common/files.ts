import * as glob from '@actions/glob';

export async function findFiles(searchPaths: string[]) {
    const globber = await glob.create(searchPaths.join('\n'));
    return await globber.glob();
}

export async function findFile(searchPath: string) {
    const paths = await findFiles([searchPath]);
    return paths[0];
}

export async function findRelativePath(searchPath: string) {
    const globber = await glob.create(searchPath);
    const searchRoot = globber.getSearchPaths()[0];
    const absolutePaths = await globber.glob();
    return absolutePaths[0] && absolutePaths[0].slice(searchRoot.length + 1);
}
