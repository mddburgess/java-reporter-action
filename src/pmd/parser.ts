import { decode } from "html-entities";
import { parseAttrs } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";
import ReportParser from "../common/parser";
import PmdReport, { PmdViolation } from "./types";

export default class PmdParser extends ReportParser<PmdReport> {
  private filePath = "";
  private violation?: PmdViolation;

  constructor(reportPath: string) {
    super(
      {
        violations: [],
      },
      reportPath
    );
  }

  protected onTagOpen(tag: TagOpenNode): void {
    switch (tag.name) {
      case "file":
        this.onFileOpen(parseAttrs(tag.attrs) as FileAttrs);
        break;
      case "violation":
        this.onViolationOpen(parseAttrs(tag.attrs) as ViolationAttrs);
        break;
    }
  }

  private onFileOpen(attrs: FileAttrs) {
    this.filePath = decode(attrs.name, { level: "xml" });
  }

  private onViolationOpen(attrs: ViolationAttrs) {
    this.violation = {
      filePath: this.filePath,
      startLine: Number(attrs.beginline),
      endLine: Number(attrs.endline),
      startColumn: Number(attrs.begincolumn),
      endColumn: Number(attrs.endcolumn),
      ruleset: decode(attrs.ruleset, { level: "xml" }),
      rule: decode(attrs.rule, { level: "xml" }),
      priority: decode(attrs.priority, { level: "xml" }),
      message: "",
    };
  }

  protected onTagClose(tag: TagCloseNode): void {
    if (tag.name !== "violation" || !this.violation) {
      return;
    }

    this.violation.message = this.violation.message.trim();
    if (!this.violation.message.endsWith(".")) {
      this.violation.message += ".";
    }
    this.report.violations.push(this.violation);
    this.violation = undefined;
  }

  protected onText(tag: TextNode | CDATANode): void {
    this.violation && (this.violation.message += decode(tag.contents, { level: "xml" }));
  }
}

interface FileAttrs {
  name: string;
}

interface ViolationAttrs {
  beginline: string;
  endline: string;
  begincolumn: string;
  endcolumn: string;
  rule: string;
  ruleset: string;
  priority: string;
}
