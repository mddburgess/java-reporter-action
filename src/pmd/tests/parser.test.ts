import fs from 'fs';
import {findFile} from '../../common/files';
import {PmdParser} from '../parser';

describe('PMD parser', () => {

    it('can parse a PMD report', async () => {
        const path = await findFile('**/pmd.xml');
        const xml = await fs.promises.readFile(path);

        const pmdReport = new PmdParser().parse(xml);
        expect(pmdReport).toMatchSnapshot();
    });
});
