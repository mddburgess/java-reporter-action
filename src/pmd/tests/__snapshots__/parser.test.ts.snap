// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PMD parser can parse a PMD report 1`] = `
Object {
  "violations": Array [
    Object {
      "endColumn": 20,
      "endLine": 3,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/BestPractices.java",
      "message": "Avoid unused imports such as 'java.io.File'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 3,
    },
    Object {
      "endColumn": 41,
      "endLine": 11,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/BestPractices.java",
      "message": "Do not hard code the IP address.",
      "rule": "AvoidUsingHardCodedIP",
      "ruleset": "Best Practices",
      "startColumn": 31,
      "startLine": 11,
    },
    Object {
      "endColumn": 37,
      "endLine": 12,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/BestPractices.java",
      "message": "Avoid unused private fields such as 'unusedPrivateField'.",
      "rule": "UnusedPrivateField",
      "ruleset": "Best Practices",
      "startColumn": 20,
      "startLine": 12,
    },
    Object {
      "endColumn": 66,
      "endLine": 17,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/BestPractices.java",
      "message": "Always check the return of one of the navigation method (next,previous,first,last) of a ResultSet.",
      "rule": "CheckResultSet",
      "ruleset": "Best Practices",
      "startColumn": 9,
      "startLine": 17,
    },
    Object {
      "endColumn": 65,
      "endLine": 21,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/BestPractices.java",
      "message": "Avoid unused method parameters such as 'unusedFormalParameter'.",
      "rule": "UnusedFormalParameter",
      "ruleset": "Best Practices",
      "startColumn": 45,
      "startLine": 21,
    },
    Object {
      "endColumn": 66,
      "endLine": 21,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/BestPractices.java",
      "message": "Avoid unused private methods such as 'unusedPrivateMethod(String)'.",
      "rule": "UnusedPrivateMethod",
      "ruleset": "Best Practices",
      "startColumn": 18,
      "startLine": 21,
    },
    Object {
      "endColumn": 31,
      "endLine": 22,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/BestPractices.java",
      "message": "Avoid unused local variables such as 'unusedLocalVariable'.",
      "rule": "UnusedLocalVariable",
      "ruleset": "Best Practices",
      "startColumn": 13,
      "startLine": 22,
    },
    Object {
      "endColumn": 2,
      "endLine": 59,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Too many static imports may lead to messy code.",
      "rule": "TooManyStaticImports",
      "ruleset": "Code Style",
      "startColumn": 1,
      "startLine": 1,
    },
    Object {
      "endColumn": 24,
      "endLine": 3,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid importing anything from the package java.lang.",
      "rule": "DontImportJavaLang",
      "ruleset": "Code Style",
      "startColumn": 1,
      "startLine": 3,
    },
    Object {
      "endColumn": 24,
      "endLine": 3,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused imports such as 'java.lang.String'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 3,
    },
    Object {
      "endColumn": 24,
      "endLine": 4,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid duplicate imports such as 'java.lang.String'.",
      "rule": "DuplicateImports",
      "ruleset": "Code Style",
      "startColumn": 1,
      "startLine": 4,
    },
    Object {
      "endColumn": 24,
      "endLine": 4,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid importing anything from the package java.lang.",
      "rule": "DontImportJavaLang",
      "ruleset": "Code Style",
      "startColumn": 1,
      "startLine": 4,
    },
    Object {
      "endColumn": 64,
      "endLine": 6,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused imports such as 'java.time.format.DateTimeFormatter.BASIC_ISO_DATE'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 6,
    },
    Object {
      "endColumn": 58,
      "endLine": 7,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused imports such as 'java.time.format.DateTimeFormatter.ISO_DATE'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 7,
    },
    Object {
      "endColumn": 63,
      "endLine": 8,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused imports such as 'java.time.format.DateTimeFormatter.ISO_DATE_TIME'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 8,
    },
    Object {
      "endColumn": 64,
      "endLine": 9,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused imports such as 'java.time.format.DateTimeFormatter.ISO_LOCAL_TIME'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 9,
    },
    Object {
      "endColumn": 61,
      "endLine": 10,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused imports such as 'java.time.format.DateTimeFormatter.ISO_INSTANT'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 10,
    },
    Object {
      "endColumn": 37,
      "endLine": 12,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "No need to explicitly extend Object.",
      "rule": "ExtendsObject",
      "ruleset": "Code Style",
      "startColumn": 32,
      "startLine": 12,
    },
    Object {
      "endColumn": 9,
      "endLine": 17,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "This for loop could be simplified to a while loop.",
      "rule": "ForLoopShouldBeWhileLoop",
      "ruleset": "Code Style",
      "startColumn": 9,
      "startLine": 15,
    },
    Object {
      "endColumn": 30,
      "endLine": 21,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused local variables such as 'value'.",
      "rule": "UnusedLocalVariable",
      "ruleset": "Best Practices",
      "startColumn": 26,
      "startLine": 21,
    },
    Object {
      "endColumn": 24,
      "endLine": 21,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary use of fully qualified name 'java.lang.String' due to existing import 'java.lang.String'.",
      "rule": "UnnecessaryFullyQualifiedName",
      "ruleset": "Code Style",
      "startColumn": 9,
      "startLine": 21,
    },
    Object {
      "endColumn": 5,
      "endLine": 26,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'static' on interface 'UnnecessaryModifier': member interfaces are implicitly static.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 24,
    },
    Object {
      "endColumn": 5,
      "endLine": 30,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'static' on enum 'UnnecessaryModifierEnum': nested enums are implicitly static.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 28,
    },
    Object {
      "endColumn": 15,
      "endLine": 33,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unnecessary return statements.",
      "rule": "UnnecessaryReturn",
      "ruleset": "Code Style",
      "startColumn": 9,
      "startLine": 33,
    },
    Object {
      "endColumn": 17,
      "endLine": 37,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Useless parentheses.",
      "rule": "UselessParentheses",
      "ruleset": "Code Style",
      "startColumn": 17,
      "startLine": 37,
    },
    Object {
      "endColumn": 26,
      "endLine": 41,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Avoid unused local variables such as 'instance'.",
      "rule": "UnusedLocalVariable",
      "ruleset": "Best Practices",
      "startColumn": 19,
      "startLine": 41,
    },
    Object {
      "endColumn": 38,
      "endLine": 41,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Useless qualified this usage in the same class.",
      "rule": "UselessQualifiedThis",
      "ruleset": "Code Style",
      "startColumn": 30,
      "startLine": 41,
    },
    Object {
      "endColumn": 42,
      "endLine": 47,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifiers 'public abstract' on method 'publicAbstract': the method is declared in an interface type.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 21,
      "startLine": 47,
    },
    Object {
      "endColumn": 50,
      "endLine": 48,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifiers 'public static final' on field 'publicStaticFinal': the field is declared in an interface type.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 25,
      "startLine": 48,
    },
    Object {
      "endColumn": 44,
      "endLine": 49,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'public' on class 'PublicStaticClass': members of interface types are implicitly public.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 49,
    },
    Object {
      "endColumn": 44,
      "endLine": 49,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'static' on class 'PublicStaticClass': types nested within an interface type are implicitly static.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 49,
    },
    Object {
      "endColumn": 52,
      "endLine": 50,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'public' on interface 'PublicStaticInterface': members of interface types are implicitly public.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 50,
    },
    Object {
      "endColumn": 52,
      "endLine": 50,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'static' on interface 'PublicStaticInterface': member interfaces are implicitly static.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 50,
    },
    Object {
      "endColumn": 41,
      "endLine": 55,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifiers 'public abstract' on method 'publicAbstract': the method is declared in an annotation type.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 21,
      "startLine": 55,
    },
    Object {
      "endColumn": 50,
      "endLine": 56,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifiers 'public static final' on field 'publicStaticFinal': the field is declared in an annotation type.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 25,
      "startLine": 56,
    },
    Object {
      "endColumn": 44,
      "endLine": 57,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'public' on class 'PublicStaticClass': members of annotation types are implicitly public.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 57,
    },
    Object {
      "endColumn": 44,
      "endLine": 57,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'static' on class 'PublicStaticClass': types nested within an interface type are implicitly static.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 57,
    },
    Object {
      "endColumn": 52,
      "endLine": 58,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'public' on interface 'PublicStaticInterface': members of annotation types are implicitly public.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 58,
    },
    Object {
      "endColumn": 52,
      "endLine": 58,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/CodeStyle.java",
      "message": "Unnecessary modifier 'static' on interface 'PublicStaticInterface': member interfaces are implicitly static.",
      "rule": "UnnecessaryModifier",
      "ruleset": "Code Style",
      "startColumn": 19,
      "startLine": 58,
    },
    Object {
      "endColumn": 13,
      "endLine": 9,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Design.java",
      "message": "These nested if statements could be combined.",
      "rule": "CollapsibleIfStatements",
      "ruleset": "Design",
      "startColumn": 13,
      "startLine": 7,
    },
    Object {
      "endColumn": 43,
      "endLine": 14,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Design.java",
      "message": "Ternary operators that can be simplified with || or &amp;&amp;.",
      "rule": "SimplifiedTernary",
      "ruleset": "Design",
      "startColumn": 16,
      "startLine": 14,
    },
    Object {
      "endColumn": 5,
      "endLine": 19,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Design.java",
      "message": "Overriding method merely calls super.",
      "rule": "UselessOverridingMethod",
      "ruleset": "Design",
      "startColumn": 12,
      "startLine": 17,
    },
    Object {
      "endColumn": 37,
      "endLine": 9,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid unused imports such as 'org.example.pmd.BestPractices'.",
      "rule": "UnusedImports",
      "ruleset": "Best Practices",
      "startColumn": 1,
      "startLine": 9,
    },
    Object {
      "endColumn": 37,
      "endLine": 9,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "No need to import a type that lives in the same package.",
      "rule": "ImportFromSamePackage",
      "ruleset": "Error Prone",
      "startColumn": 1,
      "startLine": 9,
    },
    Object {
      "endColumn": 13,
      "endLine": 13,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Empty initializer was found.",
      "rule": "EmptyInitializer",
      "ruleset": "Error Prone",
      "startColumn": 12,
      "startLine": 13,
    },
    Object {
      "endColumn": 18,
      "endLine": 20,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid using a branching statement as the last in a loop.",
      "rule": "AvoidBranchingStatementAsLastInLoop",
      "ruleset": "Error Prone",
      "startColumn": 13,
      "startLine": 20,
    },
    Object {
      "endColumn": 36,
      "endLine": 25,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid creating BigDecimal with a decimal (float/double) literal. Use a String literal.",
      "rule": "AvoidDecimalLiteralsInBigDecimalConstructor",
      "ruleset": "Error Prone",
      "startColumn": 16,
      "startLine": 25,
    },
    Object {
      "endColumn": 21,
      "endLine": 29,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Using multiple unary operators may be a bug, and/or is confusing.",
      "rule": "AvoidMultipleUnaryOperators",
      "ruleset": "Error Prone",
      "startColumn": 16,
      "startLine": 29,
    },
    Object {
      "endColumn": 18,
      "endLine": 33,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Do not start a literal by 0 unless its an octal value.",
      "rule": "AvoidUsingOctalValues",
      "ruleset": "Error Prone",
      "startColumn": 16,
      "startLine": 33,
    },
    Object {
      "endColumn": 9,
      "endLine": 39,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Method call on object which may be null.",
      "rule": "BrokenNullCheck",
      "ruleset": "Error Prone",
      "startColumn": 9,
      "startLine": 37,
    },
    Object {
      "endColumn": 16,
      "endLine": 45,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Check the value returned by the skip() method of an InputStream to see if the requested number of bytes has been skipped.",
      "rule": "CheckSkipResult",
      "ruleset": "Error Prone",
      "startColumn": 9,
      "startLine": 45,
    },
    Object {
      "endColumn": 58,
      "endLine": 51,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "This usage of the Collection.toArray() method will throw a ClassCastException.",
      "rule": "ClassCastExceptionWithToArray",
      "ruleset": "Error Prone",
      "startColumn": 39,
      "startLine": 51,
    },
    Object {
      "endColumn": 18,
      "endLine": 55,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Dont use floating point for loop indices. If you must use floating point, use double.",
      "rule": "DontUseFloatTypeForLoopIndices",
      "ruleset": "Error Prone",
      "startColumn": 14,
      "startLine": 55,
    },
    Object {
      "endColumn": 9,
      "endLine": 63,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty try blocks.",
      "rule": "EmptyTryBlock",
      "ruleset": "Error Prone",
      "startColumn": 13,
      "startLine": 61,
    },
    Object {
      "endColumn": 9,
      "endLine": 65,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty catch blocks.",
      "rule": "EmptyCatchBlock",
      "ruleset": "Error Prone",
      "startColumn": 11,
      "startLine": 63,
    },
    Object {
      "endColumn": 9,
      "endLine": 67,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty finally blocks.",
      "rule": "EmptyFinallyBlock",
      "ruleset": "Error Prone",
      "startColumn": 11,
      "startLine": 65,
    },
    Object {
      "endColumn": 9,
      "endLine": 73,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty if statements.",
      "rule": "EmptyIfStmt",
      "ruleset": "Error Prone",
      "startColumn": 24,
      "startLine": 71,
    },
    Object {
      "endColumn": 10,
      "endLine": 77,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty block statements.",
      "rule": "EmptyStatementBlock",
      "ruleset": "Error Prone",
      "startColumn": 9,
      "startLine": 77,
    },
    Object {
      "endColumn": 9,
      "endLine": 81,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "An empty statement (semicolon) not part of a loop.",
      "rule": "EmptyStatementNotInLoop",
      "ruleset": "Error Prone",
      "startColumn": 9,
      "startLine": 81,
    },
    Object {
      "endColumn": 9,
      "endLine": 88,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty switch statements.",
      "rule": "EmptySwitchStatements",
      "ruleset": "Error Prone",
      "startColumn": 9,
      "startLine": 86,
    },
    Object {
      "endColumn": 9,
      "endLine": 94,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty synchronized blocks.",
      "rule": "EmptySynchronizedBlock",
      "ruleset": "Error Prone",
      "startColumn": 29,
      "startLine": 92,
    },
    Object {
      "endColumn": 9,
      "endLine": 100,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid empty while statements.",
      "rule": "EmptyWhileStmt",
      "ruleset": "Error Prone",
      "startColumn": 22,
      "startLine": 98,
    },
    Object {
      "endColumn": 13,
      "endLine": 107,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid modifying an outer loop incrementer in an inner loop for update expression.",
      "rule": "JumbledIncrementer",
      "ruleset": "Error Prone",
      "startColumn": 13,
      "startLine": 105,
    },
    Object {
      "endColumn": 48,
      "endLine": 112,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "The null check here is misplaced; if the variable 'string' is null there will be a NullPointerException.",
      "rule": "MisplacedNullCheck",
      "ruleset": "Error Prone",
      "startColumn": 43,
      "startLine": 112,
    },
    Object {
      "endColumn": 39,
      "endLine": 115,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Ensure you override both equals() and hashCode().",
      "rule": "OverrideBothEqualsAndHashcode",
      "ruleset": "Error Prone",
      "startColumn": 20,
      "startLine": 115,
    },
    Object {
      "endColumn": 25,
      "endLine": 123,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid returning from a finally block.",
      "rule": "ReturnFromFinallyBlock",
      "ruleset": "Error Prone",
      "startColumn": 13,
      "startLine": 123,
    },
    Object {
      "endColumn": 16,
      "endLine": 128,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Do not use if statements that are always true or always false.",
      "rule": "UnconditionalIfStatement",
      "ruleset": "Error Prone",
      "startColumn": 13,
      "startLine": 128,
    },
    Object {
      "endColumn": 38,
      "endLine": 135,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Avoid unnecessary temporaries when converting primitives to Strings.",
      "rule": "UnnecessaryConversionTemporary",
      "ruleset": "Error Prone",
      "startColumn": 30,
      "startLine": 135,
    },
    Object {
      "endColumn": 38,
      "endLine": 139,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "Invoke equals() on the object you've already ensured is not null.",
      "rule": "UnusedNullCheckInEquals",
      "ruleset": "Error Prone",
      "startColumn": 38,
      "startLine": 139,
    },
    Object {
      "endColumn": 14,
      "endLine": 144,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
      "message": "An operation on an Immutable object (String, BigDecimal or BigInteger) wont change the object itself.",
      "rule": "UselessOperationOnImmutable",
      "ruleset": "Error Prone",
      "startColumn": 9,
      "startLine": 144,
    },
    Object {
      "endColumn": 22,
      "endLine": 8,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Multithreading.java",
      "message": "Avoid unused local variables such as 'tg'.",
      "rule": "UnusedLocalVariable",
      "ruleset": "Best Practices",
      "startColumn": 21,
      "startLine": 8,
    },
    Object {
      "endColumn": 40,
      "endLine": 8,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Multithreading.java",
      "message": "Avoid using java.lang.ThreadGroup; it is not thread safe.",
      "rule": "AvoidThreadGroup",
      "ruleset": "Multithreading",
      "startColumn": 30,
      "startLine": 8,
    },
    Object {
      "endColumn": 26,
      "endLine": 12,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Multithreading.java",
      "message": "Dont call Thread.run() explicitly, use Thread.start().",
      "rule": "DontCallThreadRun",
      "ruleset": "Multithreading",
      "startColumn": 9,
      "startLine": 12,
    },
    Object {
      "endColumn": 5,
      "endLine": 24,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Multithreading.java",
      "message": "Double checked locking is not thread safe in Java.",
      "rule": "DoubleCheckedLocking",
      "ruleset": "Multithreading",
      "startColumn": 12,
      "startLine": 15,
    },
    Object {
      "endColumn": 34,
      "endLine": 8,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Performance.java",
      "message": "Dont create instances of already existing BigInteger and BigDecimal (ZERO, ONE, TEN).",
      "rule": "BigIntegerInstantiation",
      "ruleset": "Performance",
      "startColumn": 16,
      "startLine": 8,
    },
    Object {
      "endColumn": 34,
      "endLine": 12,
      "filePath": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/Performance.java",
      "message": "Avoid instantiating Boolean objects; reference Boolean.TRUE or Boolean.FALSE or call Boolean.valueOf() instead.",
      "rule": "BooleanInstantiation",
      "ruleset": "Performance",
      "startColumn": 16,
      "startLine": 12,
    },
  ],
}
`;
